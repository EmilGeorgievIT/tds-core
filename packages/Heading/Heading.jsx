import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'

import { typography } from '../../shared/styles/Typography/typography'
import { breakpoints } from '@tds/core-responsive/breakpoints.js'
import colors from '../../shared/utils/colors'
import safeRest from '../../shared/utils/safeRest'

import HeadingSup from './HeadingSup/HeadingSup'

const h1Styles = {
  ...typography.wordBreak,
  padding: 0,
  margin: 0,

  ...typography.helveticaNeueLight45,
  color: colors.tokens.secondary,

  fontSize: '1.75rem',
  lineHeight: 1.29,
  letterSpacing: -1.6,
  sup: {
    fontSize: '1.25rem',
    top: '-1em',
  },

  [`@media (min-width: ${breakpoints.md}px)`]: {
    ...typography.helveticaNeueThin35,
    fontSize: '2.75rem',
    lineHeight: 1.18,
    letterSpacing: 0,
    sup: {
      fontSize: '1.25rem',
      top: '-1.3em',
    },
  },
}

const h2Styles = {
  ...typography.wordBreak,
  padding: 0,
  margin: 0,

  ...typography.helveticaNeueLight45,
  color: colors.tokens.secondary,

  fontSize: '1.5rem',
  lineHeight: 1.33,
  letterSpacing: -0.7,
  sup: {
    fontSize: '1rem',
    top: '-0.8em',
  },

  [`@media (min-width: ${breakpoints.md}px)`]: {
    ...typography.helveticaNeueThin35,
    fontSize: '1.75rem',
    lineHeight: 1.29,
    letterSpacing: -0.8,
    sup: {
      fontSize: '1rem',
      top: '-0.7em',
    },
  },
}

const smallHeading = {
  ...typography.wordBreak,
  ...typography.helveticaNeueMedium65,
  padding: 0,
  margin: 0,
  letterSpacing: -0.6,
  color: colors.typography.text,
  sup: {
    fontSize: '0.875rem',
    top: '-0.5em',
  },
}

const h3Styles = {
  ...smallHeading,
  fontSize: '1.25rem',
  lineHeight: 1.4,
}

const h4Styles = {
  ...smallHeading,
  fontSize: '1rem',
  lineHeight: 1.25,
}

const StyledHeadingTag = styled.span(props => ({
  ...(props.level === 'h1' && h1Styles),
  ...(props.level === 'h2' && h2Styles),
  ...(props.level === 'h3' && h3Styles),
  ...(props.level === 'h4' && h4Styles),
  ...(props.invert && { color: colors.white }),
}))

/**
 * Page headings. Renders an HTML `<h1-h4>` element.
 *
 * @version ./package.json
 */
const Heading = ({ level, tag = level, invert, children, ...rest }) => {
  return (
    <StyledHeadingTag {...safeRest(rest)} as={tag} level={level} invert={invert}>
      {children}
    </StyledHeadingTag>
  )
}

Heading.propTypes = {
  /**
   * The visual level of the heading. If `tag` is not specified, then `level` determines what HTML element to render.
   */
  level: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4']).isRequired,
  /**
   * The semantic level of the heading. Renders the specified HTML element, otherwise it matches `level`.
   */
  tag: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4']),
  /**
   * Invert the text color to appear light on dark backgrounds.
   */
  invert: PropTypes.bool,
  /**
   * The content. Can be text, other components, or HTML elements.
   */
  children: PropTypes.node.isRequired,
}
Heading.defaultProps = {
  invert: false,
  tag: undefined,
}

Heading.Sup = HeadingSup

export default Heading
