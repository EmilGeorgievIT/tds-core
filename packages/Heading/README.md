# TDS Core: Heading

Page headings. Renders an HTML `<h1-h4>` element.
