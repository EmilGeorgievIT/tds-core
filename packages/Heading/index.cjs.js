require('./dist/index.css')
const Heading = require('./dist/index.cjs')

module.exports = Heading
