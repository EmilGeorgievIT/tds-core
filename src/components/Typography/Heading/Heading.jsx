import React from 'react'
import PropTypes from 'prop-types'

import safeRest from '../../../safeRest'

import styles from './Heading.modules.scss'

const getColorClassName = level => (
  level === 'h1' || level === 'h2' ? styles.secondary : styles.default
)

const getClassName = (level, invert) => {
  const colorClassName = invert ? styles.inverted : getColorClassName(level)

  return `${styles[level]} ${colorClassName}`
}

const Heading = ({ level, invert, children, ...rest }) => (
  React.createElement(
    level,
    {
      ...safeRest(rest),
      className: getClassName(level, invert)
    },
    children
  )
)
Heading.propTypes = {
  level: PropTypes.oneOf([
    'h1', 'h2', 'h3', 'h4'
  ]).isRequired,
  invert: PropTypes.bool,
  children: PropTypes.node.isRequired
}
Heading.defaultProps = {
  level: 'h1',
  invert: false
}

export default Heading
