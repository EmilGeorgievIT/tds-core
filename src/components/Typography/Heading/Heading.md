### Minimal usage

```
<Heading>Great Deals</Heading>
```

### Defining sizes

```
<div>
  <Heading level="h1">Great Deals</Heading>
  <Heading level="h2">Great Deals</Heading>
  <Heading level="h3">Great Deals</Heading>
  <Heading level="h4">Great Deals</Heading>
</div>
```

### Invert on dark colours

```
const PurpleBlock = require('../../__docs__/PurpleBlock').default;

<PurpleBlock>
  <Heading invert>Great Deals</Heading>
</PurpleBlock>
```
