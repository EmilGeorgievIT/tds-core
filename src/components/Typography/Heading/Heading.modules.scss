@import '../../../scss/settings/colours';
@import '../../../scss/settings/responsive-grid';
@import '../../../scss/settings/typography';
@import '../../../scss/utility/responsive';

.largeHeading {
  composes: noSpacing from '../../Spacing.modules.scss';

  @include helvetica-neue-light-45;
}

.smallHeading {
  composes: noSpacing from '../../Spacing.modules.scss';

  @include helvetica-neue-medium-65;

  letter-spacing: -.6px;
}

.inverted {
  color: $color-white;
}

.secondary {
  color: $color-secondary;
}

.default {
  color: $color-text;
}

.h1 {
  composes: largeHeading;

  font-size: 1.75rem;
  line-height: 1.29; // 36px
  letter-spacing: -1.6px;

  @include from-breakpoint(medium) {
    font-weight: 300;
    font-size: 2.75rem;
    line-height: 1.18; // 52px
    letter-spacing: 0;
  }
}

.h2 {
  composes: largeHeading;

  font-size: 1.5rem;
  line-height: 1.33; // 30px
  letter-spacing: -.4px;

  @include from-breakpoint(medium) {
    font-size: 1.75rem;
    line-height: 1.29; // 36px
    letter-spacing: -.8px;
  }
}

.h3 {
  composes: smallHeading;

  font-size: 1.25rem;
  line-height: 1.4; // 28px
}

.h4 {
  composes: smallHeading;

  font-size: 1rem;
  line-height: 1.25; // 20px
}
