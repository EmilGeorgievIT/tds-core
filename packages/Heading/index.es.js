import './dist/index.css'

import Heading from './dist/index.es'

export default Heading
