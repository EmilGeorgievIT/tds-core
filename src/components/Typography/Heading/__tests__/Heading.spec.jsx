import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import Heading from '../Heading'

describe('Heading', () => {
  const doShallow = (overrides = {}) => shallow(
    <Heading {...overrides}>Go home</Heading>
  )

  it('renders', () => {
    const heading = doShallow()

    expect(toJson(heading)).toMatchSnapshot()
  })

  it('renders text', () => {
    const heading = doShallow()

    expect(heading).toHaveText('Go home')
  })

  it('renders a heading in four levels', () => {
    let heading = doShallow({ level: 'h1' })
    expect(heading).toHaveTagName('h1')

    heading = doShallow({ level: 'h2' })
    expect(heading).toHaveTagName('h2')

    heading = doShallow({ level: 'h3' })
    expect(heading).toHaveTagName('h3')

    heading = doShallow({ level: 'h4' })
    expect(heading).toHaveTagName('h4')
  })

  it('has appropriate colour', () => {
    let heading = doShallow({ invert: true })
    expect(heading).toHaveClassName('inverted')

    heading = doShallow({ level: 'h1' })
    expect(heading).toHaveClassName('secondary')

    heading = doShallow({ level: 'h2' })
    expect(heading).toHaveClassName('secondary')

    heading = doShallow({ level: 'h3' })
    expect(heading).toHaveClassName('default')

    heading = doShallow({ level: 'h4' })
    expect(heading).toHaveClassName('default')
  })

  it('passes additional attributes to heading element', () => {
    const heading = doShallow({ id: 'the-heading', tabindex: 1 })

    expect(heading).toHaveProp('id', 'the-heading')
    expect(heading).toHaveProp('tabindex', 1)
  })

  it('does not allow custom CSS', () => {
    const heading = doShallow({ className: 'my-custom-class', style: { color: 'hotpink' } })

    expect(heading).not.toHaveProp('className', 'my-custom-class')
    expect(heading).not.toHaveProp('style')
  })
})
