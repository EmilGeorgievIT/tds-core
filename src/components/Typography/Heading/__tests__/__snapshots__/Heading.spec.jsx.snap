// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading renders 1`] = `
<h1
  className="h1 secondary"
>
  Go home
</h1>
`;
